lib/charms/opensearch/v0/opensearch_secrets.py

```diff
@@ -0,0 +1,300 @@
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchSecrets(Object, RelationDataStore):
+    """Encapsulating Juju3 secrets handling."""
+
+    LABEL_SEPARATOR = ":"
+
+    def __init__(self, charm, peer_relation: str):
+        Object.__init__(self, charm, peer_relation)
+        RelationDataStore.__init__(self, charm, peer_relation)
+
+        self.cached_secrets = SecretCache()
+        self._jujuversion = None
+
+        self.framework.observe(self._charm.on.secret_changed, self._on_secret_changed)
+
+    def _on_secret_changed(self, event: ActionEvent):
+        """Refresh secret and re-run corresponding actions if needed."""
+        if not event.secret.label:
+            logger.info("Secret %s has no label, ignoring it", event.secret.id)
+
+        label_parts = self.breakdown_label(event.secret.label)
+        if (
+            label_parts["application_name"] == self._charm.app.name
+            and label_parts["scope"] == Scope.APP
+            and label_parts["key"] == CertType.APP_ADMIN.val
+        ):
+            scope = Scope.APP
+        else:
+            logger.info("Secret %s was not relevant for us.", event.secret.label)
+            return
+
+        logger.debug("Secret change for %s, %s", scope, label_parts["key"])
+
+        if not self._charm.unit.is_leader():
+            self._charm.store_tls_resources(CertType.APP_ADMIN, event.secret.get_content())
+
+    @property
+    def implements_secrets(self):
+        """Property to cache resutls from a Juju call."""
+        return JujuVersion.from_environ().has_secrets
+
+    def label(self, scope: Scope, key: str) -> str:
+        """Generated keys to be used within relation data to refer to secret IDs."""
+        components = [self._charm.app.name, scope.val]
+        if scope == Scope.UNIT:
+            components.append(str(self._charm.unit_id))
+        components.append(key)
+        return self.LABEL_SEPARATOR.join(components)
+
+    def breakdown_label(self, label) -> Optional[Dict[str, str]]:
+        """Return meaningful components resolved from a secret label."""
+        components = label.split(self.LABEL_SEPARATOR)
+
+        unit_id = None
+        key = None
+        if len(components) == 3:
+            key = components[2]
+        elif len(components) == 4:
+            unit_id = int(components[2])
+            key = components[3]
+
+        if components[1] == Scope.APP.val:
+            scope = Scope.APP
+        elif components[1] == Scope.UNIT.val:
+            scope = Scope.UNIT
+
+        return {
+            "application_name": components[0],
+            "scope": scope,
+            "unit_id": unit_id,
+            "key": key,
+        }
+
+    @staticmethod
+    def _safe_obj_data(indict: Dict) -> Dict:
+        return {key: str(val) for key, val in indict.items() if val is not None}
```

Comment:
Hmmm.... This would quietly "swollow" wrong (?) value `False` as a secret value... While we do nothing for `True`.

Which is inconsistent. I think we should rather be explicit and filter out the one thing (`None`) that we want to specifically exclude.

Reply:
We should not expect booleans - but to stay consistent with the API signature of [`put(self, scope: Scope, key: str, value: Optional[Union[any]])`](https://github.com/canonical/opensearch-operator/blob/0489abeeb81aa471affaf459faea132cf51317e3/lib/charms/opensearch/v0/opensearch_secrets.py#L267C59-L267C69) - let's safeguard for now with something along the lines of: `if val is not None and str(val).strip()`

We should however in a future PR revisit the `put` method to see if it's sensible to change the signature while keeping backward compatible.